### PR TITLE
docs(reference): add expression section

### LINKED ISSUE #3796: Document the expression support
**Description**

* provide documentation how expressions should look like and where they can be used
* update the documention about the [variable access](https://docs.zeebe.io/reference/variables.html#access-variables)
* update the documention about the [conditions](https://docs.zeebe.io/reference/conditions.html)
* provide the steps to migrate existing workflows

### PR DESCRIPTION
**Pull request by saig0**
## Description

* document the usage of expressions
* mention the expression support in the related BPMN workflow sections

## Related issues

closes #3796 

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [ ] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [ ] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

File: docs/src/bpmn-workflows/multi-instance/multi-instance.md [hunk 1]
```diff
@@ -21,33 +21,35 @@ A multi-instance activity is executed either sequentially or in parallel (defaul
...
+`outputElement` is an expression that defines the output of the instance (e.g. `= result`). Usually, it [access a variable](/reference/expressions.html#access-variables) of the instance that holds the output value. If the expression only access a variable or a nested property then it is created as **local variable** of the instance. This variable should be updated with the output value, for example, by a job worker providing a variable with the name `result`. Since the variable is defined as local variable, it is not propagated to its parent scope and is only visible within the instance.
+
+When the instance is completed, the `outputElement` expression is evaluated and the result is inserted into the `outputCollection` at the same index as the `inputElement` of the `inputCollection`. So, the order of the `outputCollection` is determined and matches to the `inputCollection`, even for parallel multi-instance activities. If the `outputElement` variable is not updated then `null` is inserted instead.
 
 If the `inputCollection` is empty then an empty array is propagated as `outputCollection`.
```
Comment (korthout) [inline docs/src/bpmn-workflows/multi-instance/multi-instance.md:48]:
```suggestion
If the `inputCollection` value is empty then an empty array is propagated as `outputCollection`.
```

File: docs/src/bpmn-workflows/receive-tasks/receive-tasks.md [hunk 2]
```diff
@@ -5,18 +5,18 @@ Receive tasks are tasks which references a message. They are used to wait until
...
 
 > An alternative to receive tasks are [message intermediate catch events](/bpmn-workflows/message-events/message-events.html) which behaves the same but can be used together with event-based gateways.
 
 ## Messages
 
-A message can be referenced by one or more receive tasks. It **must** define the name of the message (e.g. `Money collected`) and the `correlationKey` variable (e.g. `orderId`).
+A message can be referenced by one or more receive events. It **must** define the name of the message (e.g. `Money collected`) and the `correlationKey` expression (e.g. `= orderId`).
 
-The `correlationKey` variable must reference a variable of the workflow instance that holds the correlation key of the message. It is read from the workflow instance on activating the receive task and must be either a `string` or a `number`.
+The `correlationKey` is an expression that usually [access a variable](/reference/expressions.html#access-variables) of the workflow instance that holds the correlation key of the message. The expression is evaluated on activating the receive task and must result either in a `string` or in a `number`.
```
Comment (korthout) [inline docs/src/bpmn-workflows/receive-tasks/receive-tasks.md:17]:
```suggestion
The `correlationKey` is an expression that usually [accesses a variable](/reference/expressions.html#access-variables) of the workflow instance that holds the correlation key of the message. The expression is evaluated on activating the receive task and must result either in a `string` or in a `number`.
```

File: docs/src/bpmn-workflows/receive-tasks/receive-tasks.md [hunk 3]
```diff
@@ -5,18 +5,18 @@ Receive tasks are tasks which references a message. They are used to wait until
...
-A message can published using one of the Zeebe clients. When the message is correlated, the receive task gets completed and the workflow instance continues. 
+
+A message can published using one of the Zeebe clients. When the message is correlated, the receive task gets completed and the workflow instance continues.
 
 > An alternative to receive tasks are [message intermediate catch events](/bpmn-workflows/message-events/message-events.html) which behaves the same but can be used together with event-based gateways.
 
 ## Messages
 
-A message can be referenced by one or more receive tasks. It **must** define the name of the message (e.g. `Money collected`) and the `correlationKey` variable (e.g. `orderId`).
+A message can be referenced by one or more receive events. It **must** define the name of the message (e.g. `Money collected`) and the `correlationKey` expression (e.g. `= orderId`).
```
Comment (korthout) [inline docs/src/bpmn-workflows/receive-tasks/receive-tasks.md:15]: You've changed the name here to receive events, but the rest of this document refers to them as receive tasks.

File: docs/src/bpmn-workflows/receive-tasks/receive-tasks.md [hunk 4]
```diff
@@ -5,18 +5,18 @@ Receive tasks are tasks which references a message. They are used to wait until
...
+The `correlationKey` is an expression that usually [access a variable](/reference/expressions.html#access-variables) of the workflow instance that holds the correlation key of the message. The expression is evaluated on activating the receive task and must result either in a `string` or in a `number`.
 
-In order to correlate a message to the receive task, the message is published with the defined name (e.g. `Money collected`) and the **value** of the `correlationKey` variable. For example, if the workflow instance has a variable `orderId` with value `"order-123"` then the message must be published with the correlation key `"order-123"`.
+In order to correlate a message to the receive task, the message is published with the defined name (e.g. `Money collected`) and the **value** of the `correlationKey` expression. For example, if the workflow instance has a variable `orderId` with value `"order-123"` then the message must be published with the correlation key `"order-123"`.
 
 ## Variable Mappings
```
Comment (korthout) [inline docs/src/bpmn-workflows/receive-tasks/receive-tasks.md:21]: Also please add the `expressions.html` page to the references of this page

File: docs/src/bpmn-workflows/timer-events/timer-events.md [hunk 5]
```diff
@@ -22,23 +22,30 @@ When an intermediate timer catch event is entered then a corresponding timer is
...
 
-Timers must be defined by providing either a date, a duration, or a cycle. 
+Timers must be defined by providing either a date, a duration, or a cycle.
+
+A timer can be defined either as a static value (e.g. `PT3D`) or as an [expression](/reference/expressions.html). There are two common ways for using an expression:
+
+* [access a variable](/reference/expressions.html#access-variables) (e.g. `= remainingTime`)
+* [using temporal values](/reference/expressions.html#temporal-expressions) (e.g. `= date and time(expirationDate) - date and time(creationDate)`)
+
+The expression is evaluated on activating the timer catch event. It must result either in a `string` that has the same ISO 8601 format as the static value or an equivalent temporal value (i.e. a date-time, a duration, or a cycle).
```
Comment (korthout) [inline docs/src/bpmn-workflows/timer-events/timer-events.md:36]: This timers paragraphs is not necessarily only for timer catch events. I think we should also mention when a timer start event expression is evaluated.

File: docs/src/reference/variables.md [hunk 6]
```diff
@@ -98,35 +57,45 @@ In order to deactivate the variable propagation, the variables are set as *local
 
 Input/output variable mappings can be used to create new variables or customize how variables are merged into the workflow instance.
 
-Variable mappings are defined in the workflow as extension elements under `ioMapping`. Every variable mapping has a `source` and a `target` expression. The *source* expression defines where the value is copied from. The *target* expression defines where the value is copied to. The expressions reference a variable by its name or a nested property of a variable.
+Variable mappings are defined in the workflow as extension elements under `ioMapping`. Every variable mapping has a `source` and a `target` expression.
 
-If a variable or a nested property of a *target* expression doesn't exist then it is created. But if a variable or a nested property of a *source* expression doesn't exist then an [incident](/reference/incidents.html) is created.
+The `source` expression defines the **value** of the mapping. Usually, it [access a variable](/reference/expressions.html#access-variables) of the workflow instance that holds the value. If the variable or the nested property doesn't exist then an [incident](/reference/incidents.html) is created.
```
Comment (korthout) [inline docs/src/reference/variables.md:62]:
```suggestion
The `source` expression defines the **value** of the mapping. Usually, it [accesses a variable](/reference/expressions.html#access-variables) of the workflow instance that holds the value. If the variable or the nested property doesn't exist then an [incident](/reference/incidents.html) is created.
```

File: docs/src/reference/expressions.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+Additionally, the following attributes of BPMN elements can define an expression **optionally** instead of a static value:
+* Timer Catch Event: [timer definition](/bpmn-workflows/timer-events/timer-events.html#timers)
+* Service Task: [job type](/bpmn-workflows/service-tasks/service-tasks.html#task-definition), [job retries](/bpmn-workflows/service-tasks/service-tasks.html#task-definition)
+* Call Activity: [process id](/bpmn-workflows/call-activities/call-activities.html#defining-the-called-workflow)
+
+## Expressions vs. Static Values
+
+Some attributes of BPMN elements, like the timer definition of a timer catch event, can be defined either as a static value (e.g. `PT2H`) or as an expression (e.g. `= remaingTime`).
+
+The value is identified as an expression if it starts with an **equal sign** `=` (i.e. the expression prefix). The text behind the equal sign is the actual expression. For example, `= remaingTime` defines the expression `remaingTime` that access a variable with the name `remaingTime`.
```
Comment (korthout) [inline docs/src/reference/expressions.md:20]:
```suggestion
The value is identified as an expression if it starts with an **equal sign** `=` (i.e. the expression prefix). The text behind the equal sign is the actual expression. For example, `= remaingTime` defines the expression `remaingTime` that accesses a variable with the name `remaingTime`.
```

File: docs/src/reference/expressions.md [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+If the value doesn't have the prefix then it is used as static value. A static value is used either as a string (e.g. job type) or as a number (e.g. job retries). A string value must not be enclosed in quotes.
+
+Note that an expression can also define a static value by using literals (e.g. `"foo"`, `21`, `true`, `[1,2,3]`, `{x: 22}`, etc.).
+
+## The Expression Language
+
+An expression is written in **FEEL** (Friendly Enough Expression Language). FEEL is part of the OMG's DMN (Decision Model and Notation) specification. It is designed to have the following properties:
```
Comment (korthout) [inline docs/src/reference/expressions.md:28]: Can we refer here to an external link of the DMN and FEEL spec?

File: docs/src/reference/expressions.md [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+```
+
+Any value can be transformed into a string value using the `string()` function.
+
+```feel
+"order-" + string(orderId)
+// "order-123"
+```
+
+More functions for string values are available as [built-in functions](https://camunda.github.io/feel-scala/feel-built-in-functions#string-functions) (e.g. contains, matches, etc.).
```
Comment (korthout) [inline docs/src/reference/expressions.md:151]: Do we have a way to check that these external links are still working? They might break without us knowing.

**Review by korthout — CHANGES_REQUESTED** [review]
@saig0 Impressive documentation changes. I only have some minor change requests here and there. Please have a look.

File: docs/src/bpmn-workflows/exclusive-gateways/exclusive-gateways.md [hunk 10]
```diff
@@ -3,27 +3,31 @@
...
-A `conditionExpression` defines when a flow is taken. The expression can access the workflow instance variables and compare them with literals or other variables. Comparisons can be combined with logical operators.
+A `conditionExpression` defines when a flow is taken. It is a [boolean expression](/reference/expressions.html#boolean-expressions) that can access the workflow instance variables and compare them with literals or other variables. The condition is fulfilled when the expression returns `true`.
+
+Multiple boolean values or comparisons can be combined as disjunction (`and`) or conjunction (`or`).
 
 For example:
 
-```js
-totalPrice > 100
+```feel
```
Comment (saig0) [inline docs/src/bpmn-workflows/exclusive-gateways/exclusive-gateways.md:23]: No :sweat_smile: 

I added a simple syntax highlighting. Please have a look at `docs/highlight.feel.js`.

File: docs/src/bpmn-workflows/message-events/message-events.md [hunk 11]
```diff
@@ -30,15 +30,15 @@ When an intermediate message catch event is entered then a corresponding message
...
 
 ## Messages
 
-A message can be referenced by one or more message events. It **must** define the name of the message (e.g. `Money collected`) and the `correlationKey` variable (e.g. `orderId`), except it is only referenced by message start events.
+A message can be referenced by one or more message events. It **must** define the name of the message (e.g. `Money collected`) and the `correlationKey` expression (e.g. `= orderId`), except it is only referenced by message start events.
 
-The `correlationKey` variable must reference a variable of the workflow instance that holds the correlation key of the message. It is read from the workflow instance on activating the message event and must be either a `string` or a `number`.
+The `correlationKey` is an expression that usually [access a variable](/reference/expressions.html#access-variables) of the workflow instance that holds the correlation key of the message. The expression is evaluated on activating the message event and must result either in a `string` or in a `number`.
```
[the remark this answers is not in the document]
Comment (saig0) [reply docs/src/bpmn-workflows/message-events/message-events.md:39]: Where is the evaluation time not mentioned?

[the remark this answers is not in the document]
Comment (saig0) [reply docs/src/bpmn-workflows/message-events/message-events.md:39]: An example is in the previous sentence:
> ... and the `correlationKey` expression (e.g. `= orderId`) ...

Should we add more examples? Any suggestion?

File: docs/src/bpmn-workflows/timer-events/timer-events.md (hunk 5 above)
Comment (saig0) [reply docs/src/bpmn-workflows/timer-events/timer-events.md:36]: Good point.

File: docs/src/reference/expressions.md (hunk 8 above)
Comment (saig0) [reply docs/src/reference/expressions.md:28]: There is a reference to the DMN spec on the bottom. It links to the PDF.

File: docs/src/reference/expressions.md (hunk 9 above)
Comment (saig0) [reply docs/src/reference/expressions.md:151]: Yes. The links are checks in the CI build process.

**Comment by saig0** [comment]
@korthout, thank you for the fast review and your hints! Please have a second look :)

File: docs/src/reference/expressions.md (hunk 9 above)
Comment (korthout) [reply docs/src/reference/expressions.md:151]: That's awesome 👍

**Review by korthout — APPROVED** [review]
@saig0 LGTM

**Comment by saig0** [comment]
bors retry

**Pull request merged**